lib/matplotlib/cm.py

```diff
@@ -65,12 +66,54 @@ def _gen_cmap_d():
     # Generate reversed cmaps.
     for cmap in list(cmap_d.values()):
         rmap = cmap.reversed()
+        cmap._global = True
+        rmap._global = True
         cmap_d[rmap.name] = rmap
     return cmap_d
 
 
-cmap_d = _gen_cmap_d()
-locals().update(cmap_d)
+class _DeprecatedCmapDictWrapper(MutableMapping):
+    """Dictionary mapping for deprecated _cmap_d access."""
+
+    def __init__(self, cmap_registry):
+        self._cmap_registry = cmap_registry
+
+    def __delitem__(self, key):
+        self._warn_deprecated()
+        self._cmap_registry.__delitem__(key)
+
+    def __getitem__(self, key):
+        self._warn_deprecated()
+        return self._cmap_registry.__getitem__(key)
+
+    def __iter__(self):
```

Comment:
This does not warn for `list(mpl.cm.cmap_d)` but does warn for `list(mlp.cm.cmap_d.items())`.   Was this an oversight that `__iter__` does not warn or a decision that because it is only giving a list of keys that could then go to `get_cmap` it should not wark?

Reply:
Oversight, the iter still warns because of the getitem, but I didn't think of people just listing the colormaps and using it manually themselves elsewhere. I just added it here as well.

Reply:
same for `__len__` below?   Sorry for the thousand paper cuts at the end...

Reply:
No worries, easy enough to fix. Should have just done it myself ;)